src/libraries/System.Threading.Channels/src/System/Threading/Channels/ChannelUtilities.cs

```diff
@@ -66,60 +66,180 @@ internal static ValueTask<T> GetInvalidCompletionValueTask<T>(Exception error)
             return new ValueTask<T>(t);
         }
 
-        internal static void QueueWaiter(ref AsyncOperation<bool>? tail, AsyncOperation<bool> waiter)
+        /// <summary>Dequeues an operation from the circular doubly-linked list referenced by <paramref name="head"/>.</summary>
+        /// <param name="head">The head of the list.</param>
+        /// <param name="op">The dequeued operation.</param>
+        /// <returns>true if an operation could be dequeued; otherwise, false.</returns>
+        internal static bool TryDequeue<TAsyncOp>(ref TAsyncOp? head, [NotNullWhen(true)] out TAsyncOp? op)
```

Comment:
This method in invoked already in a thread-safe manner?

Reply:
Yes